### PR TITLE
improve stackdriver metric type

### PR DESCRIPTION
**Pull request by cyriltovena**
As per the documentation https://godoc.org/contrib.go.opencensus.io/exporter/stackdriver

This will improve the metric typing and should now appears as a gke container in stackdriver interface.

If someone has time to test it and let me know if everything is correct, that would be awesome.

### CODE REVIEW

**Comment by cyriltovena** [comment]
Why the tag approved?

**Comment by markmandel** [comment]
I think prow approves since you are an approver? @roberthbailey ?

**Comment by roberthbailey** [comment]
> Why the tag approved?

Yes, that is from prow. If you weren't an approver, the tag could be added by an approver typing `/approve` in a PR comment or by approving via the github code review flow. 

Right now it doesn't mean anything, but if we decide to switch to tide for automatic merges then it'll be required for a PR to be merged.

File: pkg/metrics/exporter.go [hunk 1]
```diff
@@ -49,19 +54,48 @@ func RegisterPrometheusExporter(registry *prom.Registry) (http.Handler, error) {
 // RegisterStackdriverExporter register a Stackdriver exporter to OpenCensus.
 // It will add Agones metrics into Stackdriver on Google Cloud.
 func RegisterStackdriverExporter(projectID string) (sd *stackdriver.Exporter, err error) {
-	// Default project will be used
+	instanceID, err := metadata.InstanceID()
```
Comment (roberthbailey) [inline pkg/metrics/exporter.go]: Does this code only run on GCP?

If not, I wonder if we should allow a user to customize the instance / zone / cluster-name values via environment variables for aggregation purposes.

File: pkg/metrics/exporter.go [hunk 2]
```diff
@@ -49,19 +54,48 @@ func RegisterPrometheusExporter(registry *prom.Registry) (http.Handler, error) {
...
+		},
 	})
 	if err != nil {
 		return
 	}
 
 	// Register it as a metrics exporter
 	view.RegisterExporter(sd)
-	return
+	return sd, nil
```
Comment (roberthbailey) [inline pkg/metrics/exporter.go]: It's confusing to have some return paths use the blank return and have others pass the return values. I think the function is long enough that you should always return the explicit values and remove the named return values from the function declaration but that's just my personal preference.

File: pkg/metrics/exporter.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/metrics/exporter.go]: No technically you can run this outside of GCP. However if you do you won't be able to correlate with existing GCP workload because `instance_id`,`zone` and `cluster_name` won't be valid which makes sense.

I think we should allow users to be able to set `DefaultMonitoringLabels *Labels` via configuration, so they can tag their custom workload, but it wasn't there already so that can be done in another PR.

File: pkg/metrics/exporter.go (hunk 2 above)
Comment (cyriltovena) [reply pkg/metrics/exporter.go]: yes I agree, I wasn't also a fan of that naked return and hesitant to change it, now I'm convinced.

File: pkg/metrics/exporter.go (hunk 1 above)
Comment (cyriltovena) [reply pkg/metrics/exporter.go]: ok let's do this.

File: pkg/metrics/exporter.go [hunk 3]
```diff
@@ -49,19 +54,48 @@ func RegisterPrometheusExporter(registry *prom.Registry) (http.Handler, error) {
...
+		logger.WithError(err).Warn("error getting cluster-name")
+		clusterName = unknown
+	}
+
 	sd, err = stackdriver.NewExporter(stackdriver.Options{
 		ProjectID: projectID,
 		// MetricPrefix helps uniquely identify your metrics.
 		MetricPrefix: "agones",
+		Resource: &monitoredres.MonitoredResource{
+			Type: "container",
```
Comment (ano12ak) [inline pkg/metrics/exporter.go]: According to the Stackdriver document, I think "Type" should be "gke_container" or "k8s_container".
https://cloud.google.com/monitoring/custom-metrics/creating-metrics#monitored_resources_for_custom_metrics

(I tried to run this build then I got following error.
2019/11/06 12:11:41 Failed to export to Stackdriver: rpc error: code = InvalidArgument desc = One or more TimeSeries could not be written: Unrecognized resource type: container.: timeSeries[0-17])

Comment (cyriltovena) [reply pkg/metrics/exporter.go]: I used the driver documentation. But yeah I think you're right.

**Comment by roberthbailey** [comment]
@cyriltovena - is this ready for another review pass?

File: pkg/metrics/exporter.go (hunk 3 above)
Comment (cyriltovena) [reply pkg/metrics/exporter.go]: changed, let me know.

**Comment by cyriltovena** [comment]
Yes !

File: pkg/metrics/util_test.go [hunk 4]
```diff
@@ -291,3 +293,57 @@ agones_gameservers_node_count_count 3
 agones_nodes_count{empty="false"} 2
 agones_nodes_count{empty="true"} 1
 `
+
+func Test_parseLabels(t *testing.T) {
+	tests := []struct {
```
Comment (roberthbailey) [inline pkg/metrics/util_test.go]: Another way you could write this (to more easily explore edge cases that should create errors) is to have two sets of tests, one that should produce errors and another that should work. I think the set of tests that should work is relatively small (you have 3 right now, which mostly covers it). But the set of inputs that should cause an error should be made much longer. This is easier if it's more compact, e.g.

```
errorInputs = []string{
  "a=b,",
  "a=b,c",
  ...
}
```

since the want is always nil and the wantErr is always true for these you can collapse them into a shorter list form, and then expand the list.

What about testing inputs with spaces added in various places? Or extra equals signs? Should non-ascii characters be accepted? Does stack driver impose any constraints on what characters can be in a key? And so on.

File: pkg/metrics/exporter.go [hunk 5]
```diff
@@ -48,20 +53,55 @@ func RegisterPrometheusExporter(registry *prom.Registry) (http.Handler, error) {
...
+
+	sd, err := stackdriver.NewExporter(stackdriver.Options{
 		ProjectID: projectID,
 		// MetricPrefix helps uniquely identify your metrics.
 		MetricPrefix: "agones",
+		Resource: &monitoredres.MonitoredResource{
+			Type: "k8s_container",
+			Labels: map[string]string{
+				"project_id":   projectID,
+				"instance_id":  instanceID,
```
Comment (roberthbailey) [inline pkg/metrics/exporter.go]: Do these have to be set? Or can they be left blank if we can't read from the metadata server?

**Comment by markmandel** [comment]
```
========================================================================
site/docs/installation/upgrading/index.html
  Non-OK status: 404 --- site/docs/installation/upgrading/index.html --> https://helm.sh/docs/helm/#helm-upgrade
========================================================================
```

Rebase / Upgrade to master - should solve that problem.

**Comment by markmandel** [comment]
@roberthbailey anything left on this before we can merge?

File: site/content/en/docs/Installation/helm.md [hunk 6]
```diff
@@ -36,8 +36,12 @@ _We recommend to install Agones in its own namespaces (like `agones-system` as s
 you can use the helm `--namespace` parameter to specify a different namespace._
 
 When running in production, Agones should be scheduled on a dedicated pool of nodes, distinct from where Game Servers are scheduled for better isolation and resiliency. By default Agones prefers to be scheduled on nodes labeled with `agones.dev/agones-system=true` and tolerates node taint `agones.dev/agones-system=true:NoExecute`. If no dedicated nodes are available, Agones will
+<<<<<<< HEAD
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/helm.md]: Need to resolve this merge conflict.

Comment (cyriltovena) [reply site/content/en/docs/Installation/helm.md]: Should be resolved.

Comment (markmandel) [reply site/content/en/docs/Installation/helm.md]: @roberthbailey is this good to go now?

(I'm on a push to close out PRs before EOY :smiley:)

File: pkg/metrics/exporter.go [hunk 7]
```diff
@@ -48,20 +52,30 @@ func RegisterPrometheusExporter(registry *prom.Registry) (http.Handler, error) {
 
 // RegisterStackdriverExporter register a Stackdriver exporter to OpenCensus.
 // It will add Agones metrics into Stackdriver on Google Cloud.
-func RegisterStackdriverExporter(projectID string) (sd *stackdriver.Exporter, err error) {
-	// Default project will be used
-	sd, err = stackdriver.NewExporter(stackdriver.Options{
+func RegisterStackdriverExporter(projectID string, defaultLabels string) (*stackdriver.Exporter, error) {
+	monitoredRes, err := getMonitoredResource(projectID)
+	if err != nil {
+		logger.WithError(err).Warn("error discovering monitored resource")
```
Comment (roberthbailey) [inline pkg/metrics/exporter.go]: Is it ok if monitoredRes is nil later on or does this need to be followed by a `return nil, err` line?

Comment (cyriltovena) [reply pkg/metrics/exporter.go]: no it's ok this is a best effort.

Comment (cyriltovena) [reply pkg/metrics/exporter.go]: by default it is nil.

**Comment by cyriltovena** [comment]
> New changes are detected. LGTM label has been removed.

just rebased buddy ! :)
